### ISSUE TITLE
Require section not merged when using branch alias

### ISSUE THREAD

**Issue by astax-t** [wikimedia/composer-merge-plugin#62]
Related to https://github.com/wikimedia/composer-merge-plugin/issues/26 and https://github.com/wikimedia/composer-merge-plugin/pull/39

The patch has fixed fatal error, but merge still doesn't work properly. Dependencies from the included files are not loaded by composer.

To reproduce:

  * Create empty git repository: `git init`

  * Create these two files in the same folder:

composer.json
```
{
    "require": {
        "wikimedia/composer-merge-plugin": "dev-master"
    },
    "extra": {
        "merge-plugin": {
            "include": [
                "composer.local.json"
            ]
        },
        "branch-alias": {
            "dev-master": "8.0.x-dev"
        }
    }
}
```

composer.local.json
```
{
    "require": {
        "psr/log": "~1.0"
    }
}
```

  * Add and commit:
```
git add -A
git commit -m "test"
```

  * Run composer: `composer update`. Output:
```
$ composer update
Loading composer repositories with package information
Updating dependencies (including require-dev)
  - Installing wikimedia/composer-merge-plugin (dev-master d3d9500)
    Cloning d3d9500b937b0cd2528fc597f8674759ea0adc67

Writing lock file
Generating autoload files
Loading composer repositories with package information
Updating dependencies (including require-dev)
Writing lock file
Generating autoload files
```

Note, "psr/log" package is not installed.

Now edit composer.json, remove "branch-alias" section and run composer again - this time "psr/log" IS installed.

If you restore branch alias in composer.json and run composer again, "psr/log" will be removed.

**Comment by bd808** [comment]
Testing with setup from https://github.com/wikimedia/composer-merge-plugin/issues/62#issue-104504807

```
$ composer -vvv update
Reading ./composer.json
Loading config file /Users/bd808/.composer/config.json
Loading config file /Users/bd808/.composer/auth.json
Loading config file ./composer.json
Executing command (CWD): git describe --exact-match --tags
Executing command (CWD): git branch --no-color --no-abbrev -v
Reading /Users/bd808/.composer/composer.json
Loading config file /Users/bd808/.composer/config.json
Loading config file /Users/bd808/.composer/auth.json
Loading config file /Users/bd808/.composer/composer.json
Loading config file /Users/bd808/.composer/auth.json
Executing command (CWD): git describe --exact-match --tags
Executing command (CWD): git branch --no-color --no-abbrev -v
Loading composer repositories with package information
Downloading https://packagist.org/packages.json
Writing /Users/bd808/.composer/cache/repo/https---packagist.org/packages.json into cache
Updating dependencies (including require-dev)
Reading /Users/bd808/.composer/cache/repo/https---packagist.org/p-provider-2013.json from cache
Reading /Users/bd808/.composer/cache/repo/https---packagist.org/p-provider-2014.json from cache
Reading /Users/bd808/.composer/cache/repo/https---packagist.org/p-provider-2014-10.json from cache
Reading /Users/bd808/.composer/cache/repo/https---packagist.org/p-provider-2015-01.json from cache
Reading /Users/bd808/.composer/cache/repo/https---packagist.org/p-provider-2015-04.json from cache
Downloading http://packagist.org/p/provider-2015-07$aa39b4ae21afbeefa5a1eff76fb6f550bfd560daff43c55bb2768b65915bc1e6.json
Writing /Users/bd808/.composer/cache/repo/https---packagist.org/p-provider-2015-07.json into cache
Reading /Users/bd808/.composer/cache/repo/https---packagist.org/p-provider-archived.json from cache
Downloading http://packagist.org/p/provider-latest$eefcb9f902932b7d49881ddf65ea324443720c0471f402800ff7003bdec6ae0c.json
Writing /Users/bd808/.composer/cache/repo/https---packagist.org/p-provider-latest.json into cache
Reading /Users/bd808/.composer/cache/repo/https---packagist.org/provider-wikimedia$composer-merge-plugin.json from cache
Analyzed 71 packages to resolve dependencies
Analyzed 67 rules to resolve dependencies
  - Installing wikimedia/composer-merge-plugin (dev-master d3d9500)
    Cloning d3d9500b937b0cd2528fc597f8674759ea0adc67
Executing command (CWD): git clone --no-checkout 'git://github.com/wikimedia/composer-merge-plugin.git' '/Users/bd808/projects/wmf/libs/cpm-62/vendor/wikimedia/composer-merge-plugin' && cd '/Users/bd808/projects/wmf/libs/cpm-62/vendor/wikimedia/composer-merge-plugin' && git remote add composer 'git://github.com/wikimedia/composer-merge-plugin.git' && git fetch composer
Executing command (/Users/bd808/projects/wmf/libs/cpm-62/vendor/wikimedia/composer-merge-plugin): git remote set-url --push origin 'git@github.com:wikimedia/composer-merge-plugin.git'
Executing command (/Users/bd808/projects/wmf/libs/cpm-62/vendor/wikimedia/composer-merge-plugin): git branch -r
Executing command (/Users/bd808/projects/wmf/libs/cpm-62/vendor/wikimedia/composer-merge-plugin): git checkout 'master' --
Executing command (/Users/bd808/projects/wmf/libs/cpm-62/vendor/wikimedia/composer-merge-plugin): git reset --hard 'd3d9500b937b0cd2528fc597f8674759ea0adc67' --

Loading plugin Wikimedia\Composer\MergePlugin
    REASON: Required by root: Install command rule (install wikimedia/composer-merge-plugin dev-master)

  [merge-plugin] composer-merge-plugin installed
Executing command (/Users/bd808/projects/wmf/libs/cpm-62/vendor/wikimedia/composer-merge-plugin): git log -n1 --pretty=%ct 'd3d9500b937b0cd2528fc597f8674759ea0adc67'
Writing lock file
Generating autoload files
  [merge-plugin] Loading composer.local.json...
  [merge-plugin] Merging psr/log
  [merge-plugin] Running additional update to apply merge settings
Reading ./composer.json
Loading config file /Users/bd808/.composer/config.json
Loading config file /Users/bd808/.composer/auth.json
Loading config file ./composer.json
Executing command (CWD): git describe --exact-match --tags
Executing command (CWD): git branch --no-color --no-abbrev -v
Reading /Users/bd808/.composer/composer.json
Loading config file /Users/bd808/.composer/config.json
Loading config file /Users/bd808/.composer/auth.json
Loading config file /Users/bd808/.composer/composer.json
Loading config file /Users/bd808/.composer/auth.json
Executing command (CWD): git describe --exact-match --tags
Executing command (CWD): git branch --no-color --no-abbrev -v
Loading plugin Wikimedia\Composer\MergePlugin_composer_tmp0
  [merge-plugin] Loading composer.local.json...
  [merge-plugin] Merging psr/log
Loading composer repositories with package information
Downloading https://packagist.org/packages.json
Writing /Users/bd808/.composer/cache/repo/https---packagist.org/packages.json into cache
Updating dependencies (including require-dev)
Reading /Users/bd808/.composer/cache/repo/https---packagist.org/p-provider-2013.json from cache
Reading /Users/bd808/.composer/cache/repo/https---packagist.org/p-provider-2014.json from cache
Reading /Users/bd808/.composer/cache/repo/https---packagist.org/p-provider-2014-10.json from cache
Reading /Users/bd808/.composer/cache/repo/https---packagist.org/p-provider-2015-01.json from cache
Reading /Users/bd808/.composer/cache/repo/https---packagist.org/p-provider-2015-04.json from cache
Reading /Users/bd808/.composer/cache/repo/https---packagist.org/p-provider-2015-07.json from cache
Reading /Users/bd808/.composer/cache/repo/https---packagist.org/p-provider-archived.json from cache
Reading /Users/bd808/.composer/cache/repo/https---packagist.org/p-provider-latest.json from cache
Reading /Users/bd808/.composer/cache/repo/https---packagist.org/provider-wikimedia$composer-merge-plugin.json from cache
Analyzed 73 packages to resolve dependencies
Analyzed 76 rules to resolve dependencies
  - Marking wikimedia/composer-merge-plugin (1.3.x-dev d3d9500) as uninstalled, alias of wikimedia/composer-merge-plugin (dev-master d3d9500)

Executing command (/Users/bd808/projects/wmf/libs/cpm-62/vendor/wikimedia/composer-merge-plugin): git log -n1 --pretty=%ct 'd3d9500b937b0cd2528fc597f8674759ea0adc67'
Reading ./composer.lock
Writing lock file
Generating autoload files
  [merge-plugin] Skipping duplicate composer.local.json...

$ composer show --installed
wikimedia/composer-merge-plugin dev-master d3d9500 Composer plugin to merge...
```

The " [merge-plugin] Merging psr/log" log message shows that the plugin trying to do the right thing, but as noted in https://github.com/wikimedia/composer-merge-plugin/pull/63/files#r38554857, the dependencies are being added to the unwrapped alias rather than true root package being used by Composer. This looks to have always been the case since the original attempt to fix #26.

**Comment by bd808** [comment]
Digging into this deeper in the Composer source, it looks to me like alias packages will cause us some deeper problems if we don't propose some upstream changes. Construction of a `RootAliasPackage` caches the 'requires', 'devRequires', 'conflicts', 'provides' and 'replaces' information of the aliased package after applying transformations based on the alias. The cached values for 'requires' and 'devRequires' are updated by calling the appropriate setters on the alias instance, but there are no similar setters for 'conflicts', 'provides' and 'replaces'.  For now we can probably just document this sadness but it would be nice to fix upstream.
